ref-exchange/src/stable_swap/mod.rs

```diff
@@ -0,0 +1,1092 @@
+                    .unwrap(),
+                "{}",
+                ERR69_MIN_RESERVE
+            );
+        }
+        self.burn_shares(&sender_id, prev_shares_amount, burn_shares);
+        env::log(
+            format!(
+                "LP {} removed {} shares by given tokens, and fee is {} shares",
+                sender_id, burn_shares, fee_part
+            )
+            .as_bytes(),
+        );
+
+        if fee_part > 0 {
+            // referral fee
+            if let Some(referral) = &fees.referral_id {
+                if self.shares.get(referral).is_some() {
+                    let referral_share = fee_part * fees.referral_fee as u128 / FEE_DIVISOR as u128;
+                    self.mint_shares(referral, referral_share);
+                    env::log(
+                        format!("Referral {} got {} shares", referral, referral_share).as_bytes(),
+                    );
+                }
+            }
+            // exchange fee
+            let exchange_share = fee_part * fees.exchange_fee as u128 / FEE_DIVISOR as u128;
+            self.mint_shares(&fees.exchange_id, exchange_share);
+            env::log(
+                format!("Admin {} got {} shares", &fees.exchange_id, exchange_share).as_bytes(),
+            );
+        }
+
+        burn_shares
+    }
+
+    /// Returns number of tokens in outcome, given amount.
+    /// Tokens are provided as indexes into token list for given pool.
+    fn internal_get_return(
+        &self,
+        token_in: usize,
+        amount_in: Balance,
+        token_out: usize,
+        fees: &AdminFees,
+    ) -> SwapResult {
+        // make amounts into comparable-amounts
+        let mut c_amount_in = amount_in;
+        let mut c_current_amounts = self.amounts.clone();
+        for (index, value) in self.token_decimals.iter().enumerate() {
+            let factor = 10_u128
+                .checked_pow((TARGET_DECIMAL - value) as u32)
+                .unwrap();
+            c_current_amounts[index] *= factor;
+            if index == token_in {
+                c_amount_in *= factor;
+            }
+        }
+
+        let invariant = self.get_invariant();
+
+        let mut ret = invariant
+            .swap_to(
+                token_in,
+                c_amount_in,
+                token_out,
+                &c_current_amounts,
+                &Fees::new(self.total_fee, &fees),
+            )
+            .expect(ERR70_SWAP_OUT_CALC_ERR);
+
+        let factor_x = 10_u128
+            .checked_pow((TARGET_DECIMAL - self.token_decimals[token_in]) as u32)
+            .unwrap();
+        let factor_y = 10_u128
+            .checked_pow((TARGET_DECIMAL - self.token_decimals[token_out]) as u32)
+            .unwrap();
+        ret.new_source_amount = ret.new_source_amount.checked_div(factor_x.into()).unwrap();
+
+        let total_y = (ret.new_destination_amount + ret.amount_swapped + ret.fee)
+            .checked_div(factor_y.into())
+            .unwrap();
```

Comment:
This still results in a rounding loss, which may be an issue once dealing with asset like GUSD, which has only 2 decimals.

But if we were storing `c_amounts` internally, it would address the issue.

Reply:
fixed.